File: office_convert/qpdf.py
"""qpdf streaming concat wrapper.

Implements FR-3 (merge step) and NFR-1 (no full output buffering). Spawns the
`qpdf` binary with `--empty --pages <list> -- -`, yields stdout in 64 KB blocks
through an async generator. Optional tee-to-disk for cache write.

The merged PDF is NEVER materialized in memory or on disk on the orchestrator
side beyond the optional cache tee.
"""

from __future__ import annotations

import asyncio
import logging
import os
from collections.abc import AsyncIterator
from contextlib import suppress
from pathlib import Path

from office_convert.errors import MergeError

log = logging.getLogger(__name__)

READ_BLOCK_SIZE = 65536  # 64 KB; matches typical Linux pipe-buffer behavior.


async def concat_streaming(
    chunk_paths: list[Path],
    cache_temp_path: Path | None = None,
    qpdf_binary: str = "qpdf",
) -> AsyncIterator[bytes]:
    """Stream the concatenation of chunk PDFs as bytes.

    Args:
        chunk_paths: ordered chunk PDFs to concatenate.
        cache_temp_path: optional path to tee streamed bytes into; the caller
            renames this into the cache atomically on success.
        qpdf_binary: name or path of the qpdf binary.

    Yields:
        Successive 64 KB byte blocks of the merged PDF.

    Raises:
        MergeError: if qpdf exits non-zero.
    """
    if not chunk_paths:
        raise MergeError(exit_code=-1, stderr_tail="no chunks to concatenate")

    argv = [qpdf_binary, "--empty", "--pages", *[str(p) for p in chunk_paths], "--", "-"]
    log.debug("invoking qpdf: %s", argv)

    proc = await asyncio.create_subprocess_exec(
        *argv,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
    )
    assert proc.stdout is not None
    assert proc.stderr is not None

    tee_handle = cache_temp_path.open("wb") if cache_temp_path else None
    try:
        while True:
            block = await proc.stdout.read(READ_BLOCK_SIZE)
            if not block:
                break
            if tee_handle is not None:
                tee_handle.write(block)
            yield block
        if tee_handle is not None:
            tee_handle.flush()
            os.fsync(tee_handle.fileno())
    finally:
        if tee_handle is not None:
            tee_handle.close()

    stderr_bytes = await proc.stderr.read()
    rc = await proc.wait()
    if rc != 0:
        # Clean up partial tee file
        if cache_temp_path is not None:
            with suppress(OSError):
                cache_temp_path.unlink(missing_ok=True)
        tail = stderr_bytes[-1024:].decode("utf-8", errors="replace")
        raise MergeError(exit_code=rc, stderr_tail=tail)

File: internal/qpdf/qpdf.go
// Package qpdf is the qpdf streaming-concat wrapper.
//
// Ported from office_convert/qpdf.py. Implements FR-3 (merge step) and NFR-1
// (no full output buffering). Spawns `qpdf --empty --pages <list> -- -` and
// copies stdout to the caller's writer in 64 KB blocks, with an optional
// tee-to-disk for the cache write.
//
// Python's async generator yielding 64 KB blocks becomes io.Copy semantics:
// the caller passes an io.Writer (the http.Flusher-backed response writer or a
// cache file), and we stream into it block by block.
package qpdf

import (
	"context"
	"io"
	"os"
	"os/exec"

	"github.com/opus2/office-convert-orchestrator/internal/oerrors"
)

const readBlockSize = 65536 // 64 KB; matches typical Linux pipe-buffer behavior.

// Binary is the qpdf executable name/path.
var Binary = "qpdf"

// ConcatStreaming streams the concatenation of chunkPaths into dst in 64 KB
// blocks. If cacheTempPath != "", streamed bytes are also tee'd to that file
// (the caller renames it into the cache atomically on success). Mirrors
// concat_streaming. Returns a MergeError if qpdf exits non-zero.
func ConcatStreaming(ctx context.Context, chunkPaths []string, dst io.Writer, cacheTempPath string) error {
	if len(chunkPaths) == 0 {
		return oerrors.NewMerge(-1, "no chunks to concatenate")
	}

	argv := append([]string{"--empty", "--pages"}, chunkPaths...)
	argv = append(argv, "--", "-")
	cmd := exec.Command(Binary, argv...)
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return oerrors.NewMerge(-1, "stdout pipe: "+err.Error())
	}
	stderr, err := cmd.StderrPipe()
	if err != nil {
		return oerrors.NewMerge(-1, "stderr pipe: "+err.Error())
	}
	if err := cmd.Start(); err != nil {
		return oerrors.NewMerge(-1, "spawn failed: "+err.Error())
	}

	// Drain stderr concurrently so qpdf never blocks on a full stderr pipe.
	var stderrBytes []byte
	stderrDone := make(chan struct{})
	go func() { stderrBytes, _ = io.ReadAll(stderr); close(stderrDone) }()

	var tee *os.File
	if cacheTempPath != "" {
		tee, err = os.Create(cacheTempPath)
		if err != nil {
			_ = cmd.Process.Kill()
			_ = cmd.Wait()
			<-stderrDone
			return oerrors.NewMerge(-1, "cache temp open: "+err.Error())
		}
	}

	var w io.Writer = dst
	if tee != nil {
		w = io.MultiWriter(dst, tee)
	}

	buf := make([]byte, readBlockSize)
	_, copyErr := io.CopyBuffer(w, stdout, buf)
	if tee != nil {
		_ = tee.Sync()
		_ = tee.Close()
	}

	<-stderrDone
	waitErr := cmd.Wait()
	rc := 0
	if cmd.ProcessState != nil {
		rc = cmd.ProcessState.ExitCode()
	}
	if rc != 0 || waitErr != nil || copyErr != nil {
		if cacheTempPath != "" {
			_ = os.Remove(cacheTempPath)
		}
		return oerrors.NewMerge(rc, tailStr(stderrBytes, 1024))
	}
	return nil
}

// ConcatToFile concatenates chunkPaths into a single file at outPath (no
// streaming to a caller needed). Used for sub-chunk merge after OOM
// subdivision. Mirrors orchestrator._merge_to_file.
func ConcatToFile(ctx context.Context, chunkPaths []string, outPath string) error {
	out, err := os.Create(outPath)
	if err != nil {
		return oerrors.NewMerge(-1, "open merge target: "+err.Error())
	}
	defer out.Close()
	return ConcatStreaming(ctx, chunkPaths, out, "")
}

func tailStr(b []byte, n int) string {
	if len(b) > n {
		b = b[len(b)-n:]
	}
	return string(b)
}
